external-crates/move/crates/move-compiler/src/typing/translate.rs

```diff
@@ -2104,213 +2071,651 @@ fn add_field_types<T>(
-        let msg = "This access will make a new copy of the constant. Consider binding the value to a variable first to make this copy explicit";
-        context
-            .env
-            .add_diag(diag!(TypeSafety::ImplicitConstantCopy, (loc, msg)))
+    let ExpDotted {
+        loc,
+        base,
+        base_type,
+        base_kind,
+        accessors,
+    } = ed;
+
+    let mut exp = match base_kind {
+        BaseRefKind::Owned => exp_to_borrow(context, loc, mut_, Box::new(base), base_type),
+        BaseRefKind::ImmRef | BaseRefKind::MutRef => Box::new(base),
+    };
+
+    for accessor in accessors {
+        check_mut(context, loc, exp.ty.clone(), mut_);
+        match accessor {
+            ExpDottedAccess::Field(name, ty) => {
+                let e_ = TE::Borrow(mut_, exp, name);
+                let ty = sp(loc, Type_::Ref(mut_, Box::new(ty)));
+                exp = Box::new(T::exp(ty, sp(loc, e_)));
+            }
+            ExpDottedAccess::Index {
+                index_loc,
+                syntax_methods,
+                args,
+                base_type: index_base_type,
+            } => {
+                let Some(index_methods) = syntax_methods else {
+                    assert!(context.env.has_errors());
+                    exp = make_error_exp(context, loc);
+                    break;
+                };
+                let (m, f) = if mut_ {
+                    if let Some(index_mut) = index_methods.index_mut {
+                        index_mut.target_function
+                    } else {
+                        let msg = "Could not find a mutable index 'syntax' method";
+                        context
+                            .env
+                            .add_diag(diag!(Declarations::MissingSyntaxMethod, (index_loc, msg),));
+                        exp = make_error_exp(context, index_loc);
+                        break;
+                    }
+                } else if let Some(index) = index_methods.index {
+                    index.target_function
+                } else {
+                    let msg = "Could not find an immutable index 'syntax' method";
+                    context
+                        .env
+                        .add_diag(diag!(Declarations::MissingSyntaxMethod, (index_loc, msg),));
+                    exp = make_error_exp(context, index_loc);
+                    break;
+                };
+                let sp!(argloc, mut args_) = args;
+                args_.insert(0, *exp);
+                let mut_type = sp(index_loc, Type_::Ref(mut_, Box::new(index_base_type)));
+                let (ret_ty, e_) = module_call(context, loc, m, f, None, argloc, args_);
+                let ret_ty = invariant(
+                    context,
+                    loc,
+                    || format!("Index syntax method '{m}::{f}' has the wrong return type"),
+                    mut_type,
+                    ret_ty,
+                );
```

Comment:
I'm a bit lost to why we need to check this again?

Reply:
This should be an `ice!` but the plumbing isn't quite there. This seemed as good a sanity check as any. I can remove it if you'd like.

Reply:
can you just make it a `panic` for now? I'd be fine with that since I do that in some other cases. We can clean up all of them then at some other point in time